core/matrix/sparsity.cpp

```diff
@@ -0,0 +1,207 @@
+{
+    using Dense = Dense<ValueType>;
+    this->get_executor()->run(
+        sparsity::make_spmv(this, as<Dense>(b), as<Dense>(x)));
+}
+
+
+template <typename ValueType, typename IndexType>
+void Sparsity<ValueType, IndexType>::apply_impl(const LinOp *alpha,
+                                                const LinOp *b,
+                                                const LinOp *beta,
+                                                LinOp *x) const
+{
+    using Dense = Dense<ValueType>;
+    this->get_executor()->run(sparsity::make_advanced_spmv(
+        as<Dense>(alpha), this, as<Dense>(b), as<Dense>(beta), as<Dense>(x)));
+}
+
+
+template <typename ValueType, typename IndexType>
+void Sparsity<ValueType, IndexType>::read(const mat_data &data)
+{
+    size_type nnz = 0;
+    for (const auto &elem : data.nonzeros) {
+        nnz += (elem.value != zero<ValueType>());
+    }
+    auto tmp =
+        Sparsity::create(this->get_executor()->get_master(), data.size, nnz);
+    size_type ind = 0;
+    size_type cur_ptr = 0;
+    tmp->get_row_ptrs()[0] = cur_ptr;
+    tmp->get_value()[0] = one<ValueType>();
+    for (size_type row = 0; row < data.size[0]; ++row) {
+        for (; ind < data.nonzeros.size(); ++ind) {
+            if (data.nonzeros[ind].row > row) {
+                break;
+            }
+            auto val = data.nonzeros[ind].value;
+            if (val != zero<ValueType>()) {
+                tmp->get_col_idxs()[cur_ptr] = data.nonzeros[ind].column;
+                ++cur_ptr;
+            }
+        }
+        tmp->get_row_ptrs()[row + 1] = cur_ptr;
+    }
+    tmp->move_to(this);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Sparsity<ValueType, IndexType>::write(mat_data &data) const
+{
+    std::unique_ptr<const LinOp> op{};
+    const Sparsity *tmp{};
+    if (this->get_executor()->get_master() != this->get_executor()) {
+        op = this->clone(this->get_executor()->get_master());
+        tmp = static_cast<const Sparsity *>(op.get());
+    } else {
+        tmp = this;
+    }
+
+    data = {tmp->get_size(), {}};
+
+    for (size_type row = 0; row < tmp->get_size()[0]; ++row) {
+        const auto start = tmp->row_ptrs_.get_const_data()[row];
+        const auto end = tmp->row_ptrs_.get_const_data()[row + 1];
+        for (auto i = start; i < end; ++i) {
+            const auto col = tmp->col_idxs_.get_const_data()[i];
+            const auto val = one<ValueType>();
+            data.nonzeros.emplace_back(row, col, val);
+        }
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Sparsity<ValueType, IndexType>::transpose() const
+{
+    auto exec = this->get_executor();
+    auto trans_cpy = Sparsity::create(exec, gko::transpose(this->get_size()),
+                                      this->get_num_nonzeros());
+
+    exec->run(sparsity::make_transpose(trans_cpy.get(), this));
+    return std::move(trans_cpy);
```

Comment:
You don't need the `std::move` here since `trans_cpy` is already a `std::unique_ptr<LinOp>`.
You are basically moving a `unique_ptr` to another `unique_ptr`, which is then returned
(not a lot of overhead, so it does not really matter, but I thought I would mention it here).

Reply:
I put it to signal that it is indeed a temporary pointer and is moved away. But I guess it is not necessary.

Reply:
Actually, I think I need a std::move here because it is a Sparsity type and needs to be converted into a LinOp. For example, Intel gives me an [error without the std::move](https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/303422786)

Reply:
Interesting, I did not know that.
Thanks!